src/main/java/io/confluent/copycat/hdfs/TopicPartitionWriter.java

```diff
@@ -0,0 +1,461 @@
+      writers.put(encodedPartition, writer);
+      return writer;
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private String getTempfile(String encodedPartition) {
+    String tempfile;
+    if (tempfiles.containsKey(encodedPartition)) {
+      tempfile = tempfiles.get(encodedPartition);
+    } else {
+      String directory = getDirectory(encodedPartition);
+      tempfile = FileUtils.tempFileName(url, topicsDir, directory);
+      tempfiles.put(encodedPartition, tempfile);
+    }
+    return tempfile;
+  }
+
+
+  private void createWAL() throws CopycatException {
+    if (wal == null) {
+      wal = storage.wal(topicsDir, tp);
+    }
+  }
+
+  private void applyWAL() throws CopycatException {
+    if (!recovered) {
+      wal.apply();
+    }
+  }
+
+  private void truncateWAL() throws CopycatException {
+    if (!recovered) {
+      wal.truncate();
+    }
+  }
+
+  private void resetOffsets() throws CopycatException {
+    if (!recovered) {
+      readOffset();
+      if (offset > 0) {
+        context.offset(tp, offset);
+      }
+      recovered = true;
+    }
+  }
+
+  private void writeRecord(SinkRecord record) throws IOException {
+    String encodedPartition = partitioner.encodePartition(record);
+    RecordWriter<Long, SinkRecord> writer = getWriter(record, encodedPartition);
+    writer.write(System.currentTimeMillis(), record);
+    if (offset == -1) {
+      offset = record.kafkaOffset() - 1;
+    }
+    if (!startOffsets.containsKey(encodedPartition)) {
+      startOffsets.put(encodedPartition, record.kafkaOffset() - 1);
+      offsets.put(encodedPartition, record.kafkaOffset() - 1);
+    } else {
+      offsets.put(encodedPartition, record.kafkaOffset() - 1);
+    }
+    recordCounter++;
+  }
+
+  private void closeTempfile(String encodedPartition) throws IOException {
+    if (writers.containsKey(encodedPartition)) {
+      RecordWriter writer = writers.get(encodedPartition);
+      writer.close();
+      writers.remove(encodedPartition);
+    }
+    recordCounter = 0;
+  }
+
+  private void closeTempfile() throws IOException {
+    for (String encodedPartition: tempfiles.keySet()) {
+      closeTempfile(encodedPartition);
+    }
+  }
+  private void appendToWAL(String encodedPartition) throws IOException {
+    String tempfile = tempfiles.get(encodedPartition);
+    if (appended.contains(tempfile)) {
+      return;
+    }
+    if (!startOffsets.containsKey(encodedPartition)) {
+      return;
+    }
+    long startOffset = startOffsets.get(encodedPartition) + 1;
+    long endOffset = offsets.get(encodedPartition) + 1;
+    String directory = getDirectory(encodedPartition);
+    String committedFile = FileUtils.committedFileName(url, topicsDir, directory, tp, startOffset, endOffset);
+    wal.append(tempfile, committedFile);
+    appended.add(tempfile);
+  }
+
+  private void appendToWAL() throws IOException {
```

Comment:
See above comment -- this method definitely needs some sort of adjustment to ensure exactly once delivery.